tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/service/websocket/RetryClientTest.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http:www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http:www.apache.orglicensesLICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specif ic language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.test.service.websocket;
+
+import org.ballerinalang.test.context.BallerinaTestException;
+import org.ballerinalang.test.util.websocket.client.WebSocketTestClient;
+import org.ballerinalang.test.util.websocket.server.WebSocketRemoteServer;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This Class tests the retry support of the WebSocket client and server.
+ */
+@Test(groups = {"websocket-test"})
+public class RetryClientTest extends WebSocketTestCommons {
+
+    private WebSocketRemoteServer remoteServer;
+    private static final String url =  "ws://localhost:21030";
+    private static final int port = 15300;
+    private static final String sentText = "hi all";
+
+    @Test(description = "Tests the retry function using the WebSocket client (starting the server " +
+            ", sending, and receiving text frames.  After that, restart that server and do the same.)")
+    public void testRetry() throws URISyntaxException, InterruptedException, BallerinaTestException {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        remoteServer = new WebSocketRemoteServer(port);
+        remoteServer.run();
+        WebSocketTestClient client = new WebSocketTestClient(url);
+        client.handshake();
+        client.setCountDownLatch(countDownLatch);
+        client.sendText(sentText);
+        countDownLatch.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        Assert.assertEquals(client.getTextReceived(), sentText);
+        remoteServer.stop();
+        CountDownLatch latchForRestart = new CountDownLatch(1);
+        client.setCountDownLatch(latchForRestart);
+        remoteServer.run();
+        latchForRestart.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        client.sendText(sentText);
+        latchForRestart.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
```

Comment:
you cannot await on the same latch twice

Reply:
When latch doesn't set to the client, we can use the same latch until the running remote server stop.